src/leader_election.rs

```diff
@@ -25,3 +25,342 @@ where
+        /// * `peers` - Vector that holds all the other replicas.
+        /// * `pid` -  Process identifier used to uniquely identify this instance.
+        /// * `hb_delay` -  A fixed delay that is added to the current_delay. It is measured in ticks.
+        /// * `increment_delay` - A fixed delay that is added to the current_delay. It is measured in ticks.
+        /// * `quick_timeout` -  Initiates a quick timeout.
+        /// * `initial_leader` -  Initial leader which will be elected.
+        /// * `initial_delay_factor` -  A factor used in the beginning for a shorter hb_delay.
+        pub fn with(
+            peers: Vec<u64>,
+            pid: u64,
+            hb_delay: u64,
+            increment_delay: u64,
+            quick_timeout: bool,
+            initial_leader: Option<Leader<Ballot>>,
+            initial_delay_factor: Option<u64>,
+        ) -> BallotLeaderElection {
+            let n = &peers.len() + 1;
+            let (leader, initial_ballot) = match initial_leader {
+                Some(l) => {
+                    let leader_ballot = Ballot::with(l.round.n, l.pid);
+                    let initial_ballot = if l.pid == pid {
+                        leader_ballot
+                    } else {
+                        Ballot::with(0, pid)
+                    };
+                    (Some(leader_ballot), initial_ballot)
+                }
+                None => {
+                    let initial_ballot = Ballot::with(0, pid);
+                    (None, initial_ballot)
+                }
+            };
+            BallotLeaderElection {
+                pid,
+                majority: n / 2 + 1, // +1 because peers is exclusive ourselves
+                peers,
+                hb_round: 0,
+                ballots: Vec::with_capacity(n),
+                current_ballot: initial_ballot,
+                majority_connected: true,
+                leader,
+                hb_current_delay: hb_delay,
+                hb_delay,
+                increment_delay,
+                quick_timeout,
+                initial_delay_factor: initial_delay_factor.unwrap_or(1),
+                ticks_elapsed: 0,
+                outgoing: vec![],
+            }
+        }
+
+        /// Get the current elected leader
+        pub fn get_leader(&self) -> Option<Leader<Ballot>> {
+            self.leader
+                .and_then(|ballot: Ballot| -> Option<Leader<Ballot>> {
+                    Some(Leader::with(ballot.pid, ballot))
+                })
+        }
+
+        /// tick is run by all servers to simulate the passage of time
+        /// Returns an Option with the elected leader otherwise None
+        pub fn tick(&mut self) -> Option<Leader<Ballot>> {
+            self.ticks_elapsed += 1;
+
+            if self.ticks_elapsed >= self.hb_current_delay {
+                self.ticks_elapsed = 0;
+                self.hb_timeout()
+            } else {
+                None
+            }
+        }
+
```

Comment:
We must have a public function that returns the outgoing messages. We implemented one in the integration tests. The documentation should be similar to tick() that asks the user to call it repeatedly.